File: problems/CSES/Sorting_and_Searching/factory_machines/factory_machines.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, t = map(int, input().split())
    list_k = list(map(int, input().split()))

    left = 0
    right = min(list_k) * t

    while left < right:
        mid = left + (right - left) // 2

        count = 0
        for machine in list_k:
            count += mid // machine

            if count >= t:
                break

        if count >= t:
            right = mid
        else:
            left = mid + 1

    print(left)

main()

File: problems/CSES/Sorting_and_Searching/factory_machines/factory_machines.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    ll t;
    cin >> n >> t;

    ll best_machine = 0;

    vector<ll> machines(n);
    for (int i = 0; i < n; i++) {
        ll x;
        cin >> x;

        machines[i] = x;
        best_machine = max(best_machine, x);
    }
    
    ll left = 0;
    ll right = best_machine * t;

    while (left < right) {
        ll mid = left + (right- left) / 2;

        ll count = 0;
        for (ll machine : machines) {
            count += mid / machine;

            if (count > t) break;
        }

        if (count >= t) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }

    cout << left << "\n";

    return 0;
}
